daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,188 @@
+import com.daml.lf.data.ImmArray
+
+private[lf] object NormalizeRollbacks {
+
+  type Nid = NodeId
+  type Cid = Value.ContractId
+  type TX = GenTransaction[Nid, Cid]
+  type Node = GenNode[Nid, Cid]
+  type ActNode = GenActionNode[Nid, Cid]
+
+  // Normalize a transaction so rollback nodes satisfy the normalization rules.
+  // see `makeRoll` below
+
+  def normalizeTx(txOriginal: TX): TX = {
+
+    // Here we generate fresh node-ids for the normalized transaction.
+    val ids = Iterator.from(0).map(NodeId(_)) // perhaps we could reuse orig nids
+
+    // There is no connection between the ids in the pre and post transaction.
+    // Semantically this is fine: The ids are just of tie the tree structure together.
+    // But perhaps for ease of debugging we might consider trying to reuse existing ids.
+
+    // We accumulate (imperatively!) a nodes-map for the normalized transaction
+    var nodes: Map[Nid, Node] = Map.empty
+
+    def pushNode(node: Node): Nid = {
+      val nodeId = ids.next()
+      nodes += (nodeId -> node)
+      nodeId
+    }
+
+    def pushNodes(nodes: List[Node]): List[Nid] = nodes.map(pushNode)
+
+    import Canonical.{Norm, Case, caseNorms}
+    // The normalization phase works by constructing intermediate values which are
+    // `Canonical` in the sense that only properly normalized nodes can be represented.
+    // Although this doesn't ensure correctness, one class of bugs is avoided.
+
+    // The `force*` functions move from the world of Norms into standard transactions.
+    // This is also where node-ids for the resulting normalized transaction are generated.
+
+    def forceRoll(x: Norm.Roll): Node = {
+      x match {
+        case Norm.Roll1(act) =>
+          val child = pushNode(act.node)
+          NodeRollback(children = ImmArray(List(child)))
+        case Norm.Roll2(h, m, t) =>
+          val nodes = List(h.node) ++ m ++ List(t.node)
+          val children = pushNodes(nodes)
+          NodeRollback(children = ImmArray(children))
+      }
+    }
+
+    def forceNorm(x: Norm): Node = {
+      x match {
+        case Norm.Act(node) => node
+        case roll: Norm.Roll => forceRoll(roll)
+      }
+    }
+
+    def forceNorms(xs: List[Norm]): List[Node] = xs.map(forceNorm)
+
+    // makeRoll: encodes the normalization transformation rules:
+    //   rule #1: R [ ] -> ε
+    //   rule #2: R [ R [ xs… ] , ys… ] -> R [ xs… ] , R [ ys… ]
+    //   rule #3: R [ xs… , R [ ys… ] ] ->  R [ xs… , ys… ]
+
+    def makeRoll(norms: List[Norm]): List[Norm] = {
+      caseNorms(norms) match {
+        case Case.Empty =>
+          // normalization rule #1
+          Nil
+
+        case Case.Single(roll: Norm.Roll) =>
+          // normalization rule #2 or #3
+          List(roll)
+
+        case Case.Single(act: Norm.Act) =>
+          // no rule
+          List(Norm.Roll1(act))
+
+        case Case.Multi(h: Norm.Roll, m, t) =>
+          // normalization rule #2
+          h :: makeRoll(m ++ List(t))
```

Comment:
This recursion here should also be made stack-safe.

Reply:
done